File: singleton/Logger.java
package singleton;

public class Logger {

    private static Logger instance;
    private LogLevel logLevel = LogLevel.INFO;
    private Logger(){}
    public void setLogLevel(LogLevel logLevel) {
        this.logLevel = logLevel;
    }

    public void logDebug(String message) {
        if(logLevel.ordinal() <= LogLevel.DEBUG.ordinal()) {
            System.out.println("DEBUG: " + message);
        }
    }
    public void logInfo(String message) {
        if(logLevel.ordinal() <= LogLevel.INFO.ordinal()) {
            System.out.println("INFO: " + message);
        }
    }
    public void logWarn(String message) {
        if(logLevel.ordinal() <= LogLevel.WARN.ordinal()) {
            System.out.println("WARN: " + message);
        }
    }
    public void logError(String message) {
        if(logLevel.ordinal() <= LogLevel.ERROR.ordinal()) {
            System.out.println("ERROR: " + message);
        }
    }

    public static Logger getInstance() {
        if (instance == null) {
            synchronized (Logger.class) {
                if(instance == null) {
                    instance = new Logger();
                }
            }
        }
        return instance;
    }
}

File: typescript/patterns/singleton/logger.ts
class Logger {
  private logLevel: LogLevel;
  // 👇 this approach is called eager initialization
  private static instance: Logger = new Logger();
  private constructor() {}

  static getInstance(): Logger {
    // 👇 this approach is called lazy initialization
    if (!this.instance) {
      this.instance = new Logger();
    }
    return this.instance;
  }

  setLogLevel(logLevel: LogLevel) {
    this.logLevel = logLevel;
  }

  logDebug(message: string) {
    if (this.logLevel === LogLevel.DEBUG) {
      console.log(message);
    }
  }
  logInfo(message: string) {
    if (this.logLevel === LogLevel.INFO) {
      console.log(message);
    }
  }
  logError(message: string) {
    if (this.logLevel === LogLevel.ERROR) {
      console.log(message);
    }
  }
  logWarn(message: string) {
    if (this.logLevel === LogLevel.WARN) {
      console.log(message);
    }
  }
}
